### ISSUE TITLE
Return list of string/object when using any function that takes a single item

### ISSUE THREAD

**Issue by rwemay** [BHoM/Excel_UI#164]
#### Description:

For example - GetProperty(Name) on a list should return a list of strings.

**Comment by rwemay** [comment]
Any comments or progress on this @awakeman ?

**Comment by awakeman** [comment]
I looked into this in the last milestone. should really have dumped my thoughts here, sorry. 

It's not as simple as it sounds due to the way that the `DataAccessor` and `Caller` interface works and the assumptions made due to it being designed to fit the behaviour of grasshopper/dynamo primarily.

Grasshopper &mdash; and presumably dynamo too, although i'm not as familiar there &mdash; has the concept of Item access, List access and Tree access at the platform level. As a result the BHoM plugin to grasshopper needs only to tell grasshopper what it expects and it will be run multiple times (through `SolveInstance`) and the `DataAccessor` will get each individual item and so on. Likewise with List access and Tree inputs.

This means that to solve for Excel we need to inject a layer above the `DataAccessor` and `Caller` that gets the objects from the `Project` dictionary, knows what the method wants and replicates this behaviour.

It's not impossible by any measure but it's a pretty significant piece of work and will need a deeper understanding on how to handle lists of a different lengths, mixes of lists and single items, etc.

Basically a whole grasshopper emulation layer.

**Comment by rwemay** [comment]
Thanks, apologies, not sure I follow on needing to build whole new emulation layer - isn't there a way through BHoM_UI/Excel_Toolkit to check whether it's a multi-out and then create a list accordingly? When I've stepped-into, it is stepping through the multiple outputs, but only storing the last output in Excel/data. Good if you can look into this asap as it's a massive restriction on Excel/any software we want to be returning results for Execute and Push functions (as one example). @adecler any insights from you much appreciated.

**Comment by awakeman** [comment]
for multiple outputs, yeah that should be simple, but that's #182.

**Comment by rwemay** [comment]
Thanks @awakeman ! glad #182 is simple. Apologies - wrong thread! 

This, I would say, is less of an issue provided we get the Linq methods wrapped asap. We should target Linq wrapping this quarter, maybe good for you to lean into my prototype as Excel_Toolkit is the main reason for wrapping in the short term.

**Comment by adecler** [comment]
I am not sure I understand the user case for this other than coming from a GH bias.

@rwemay , If I understood correctly, you want to select a list of objects, run a method on them and return an output for each input (assuming that method takes a single item at a time). Is that correct ?

Because, to me, the proper way to do that in Excel is to have each output cell containing the call to the method like any native excel function would.

**Comment by rwemay** [comment]
Hey @adecler , as mentioned, Excel currently returns 'List' objects in-cell. If I run a 'GetProperty' on a list of objects in a single cell, I would expect a new list (in a single cell) of the properties from each object. 

I'll raise separately the idea to return Lists into separate ranges as additional options/overrides for any method return.

**Comment by JosefTaylor** [comment]
similar to this is Excel's dynamic range functionality, whereby if you've got a range (either something you've named, or the output of an expand or explode BHoM function, or various others, you can access that array in another formula, like the example below:

Range | =Range^2 | =B1#/2
-- | -- | --
1 | 1 | 0.5
1 | 1 | 0.5
2 | 4 | 2
5 | 25 | 12.5
14 | 196 | 98
42 | 1764 | 882
132 | 17424 | 8712
429 | 184041 | 92020.5
1430 | 2044900 | 1022450

The first column is named "Range"; the second column contains exactly one formula, "=Range^2" which spills into the rest of the column, and the third column (my favorite) refers to the first cell of the second column with "#" which gets the complete spilled output of that cell. This is super handy stuff and makes excel spreadsheets considerably less error prone (since you only have one opportunity for the formula to be wrong). I'd love to be able to use "#" in BHoM formulae!

**Comment by adecler** [comment]
To summarise the discussion we just had on the Framework and UI sprint planning regarding this topic:

Excel doesn't have any functionality around lacing. So the main issue here is that users are biased by Grasshopper and trying to use Excel the same way. While implementing lacing for Excel is possible, it is a decent amount of work.

I personally believe that we should encourage people to use the right tool for the job. If people are struggling to use Excel like Grasshopper, it means the task is probably better suited for a GH script. 

As for the proper way to handle list outputs, the recommendation is to use the `Expand list` function provided in the BHoM ribbon. Also note there is a `Condense list` function mirroring that effect.

Finally, we should still provide support for people expecting lacing when using the BHoM in Excel. @IsakNaslundBh suggested a warning message recommending to use the `Expand list` function when a list is provided to an input that expects a single item.
